src/android/com/microsoftopentechnologies/intellij/serviceexplorer/Node.java

```diff
@@ -58,66 +58,25 @@
-        this(id, name, null, null, false);
+        this(id, name, null, null, false, false);
     }
 
     public Node(String id, String name, Node parent, String iconPath, boolean hasRefreshAction) {
+        this(id, name, parent, iconPath, hasRefreshAction, false);
+    }
+
+    public Node(String id, String name, Node parent, String iconPath, boolean hasRefreshAction, boolean delayActionLoading) {
         this.id = id;
         this.name = name;
         this.parent = parent;
         this.iconPath = iconPath;
+        this.hasRefreshAction = hasRefreshAction;
 
-        // add the click action handler
-        addClickActionListener(new NodeActionListener() {
-            @Override
-            public void actionPerformed(NodeActionEvent e) {
-                onNodeClick(e);
-            }
-        });
-
-        // add the other actions
-        Map<String, Class<? extends NodeActionListener>> actions = initActions();
-        if(actions != null) {
-            for(Map.Entry<String, Class<? extends NodeActionListener>> entry : actions.entrySet()) {
-                try {
-                    // get default constructor
-                    Class<? extends NodeActionListener> listenerClass = entry.getValue();
-                    Constructor constructor = listenerClass.getDeclaredConstructor(getClass());
-
-                    // create an instance passing this object as a constructor argument
-                    // since we assume that this is an inner class
-                    NodeActionListener actionListener = (NodeActionListener)constructor.newInstance(this);
-                    addAction(entry.getKey(), actionListener);
-
-                } catch (InstantiationException e) {
-                    UIHelper.showException(e.getMessage(), e);
-                } catch (IllegalAccessException e) {
-                    UIHelper.showException(e.getMessage(), e);
-                } catch (NoSuchMethodException e) {
-                    UIHelper.showException(e.getMessage(), e);
-                } catch (InvocationTargetException e) {
-                    UIHelper.showException(e.getMessage(), e);
-                }
-            }
-        }
-
-        // add the refresh node action
-        if(hasRefreshAction) {
-            addAction("Refresh", new NodeActionListener() {
-                @Override
-                public void actionPerformed(NodeActionEvent e) {
-                    Futures.addCallback(load(), new FutureCallback<List<Node>>() {
-                        @Override
-                        public void onSuccess(List<Node> nodes) {}
-
-                        @Override
-                        public void onFailure(Throwable throwable) {
-                            UIHelper.showException("An error occurred while refreshing the service.", throwable);
-                        }
-                    });
-                }
-            });
+        if (!delayActionLoading) {
+            loadActions();
```

Comment:
The only issue I see with the invokeLater approach is that we cannot say, in a deterministic way, when it's going to run. Since we delegate that responsability to the ApplicationManager in IntelliJ, although it will most probably work, I cannot foretell it wont bring us any race condition kind of issues.
I would leave the invokeLater usage only for UI, to prevent blocking during network operations or similar, but I wouldn't use it to solve an issue that could be fixed with some code refactoring.